CircuitPython_WeatherCloud/code.py

```diff
@@ -0,0 +1,119 @@
+import time
+import random
+import board
+import busio
+from digitalio import DigitalInOut
+import neopixel
+from adafruit_esp32spi import adafruit_esp32spi
+from adafruit_esp32spi import adafruit_esp32spi_wifimanager
+import adafruit_fancyled.adafruit_fancyled as fancy
+
+print("ESP32 Open Weather API demo")
+
+# Use cityname, country code where countrycode is ISO3166 format.
+# E.g. "New York, US" or "London, GB"
+LOCATION = "Manhattan, US"
+
+# Get wifi details and more from a secrets.py file
+try:
+    from secrets import secrets
+except ImportError:
+    print("WiFi secrets are kept in secrets.py, please add them there!")
+    raise
+
+# Set up where we'll be fetching data from
+DATA_SOURCE = "http://api.openweathermap.org/data/2.5/weather?q="+LOCATION
+DATA_SOURCE += "&appid="+secrets['openweather_token']
+
+# If you are using a board with pre-defined ESP32 Pins:
+esp32_cs = DigitalInOut(board.ESP_CS)
+esp32_ready = DigitalInOut(board.ESP_BUSY)
+esp32_reset = DigitalInOut(board.ESP_RESET)
+
+spi = busio.SPI(board.SCK, board.MOSI, board.MISO)
+esp = adafruit_esp32spi.ESP_SPIcontrol(spi, esp32_cs, esp32_ready, esp32_reset)
+status_light = neopixel.NeoPixel(board.NEOPIXEL, 1, brightness=0.2) # Uncomment for Most Boards
+wifi = adafruit_esp32spi_wifimanager.ESPSPI_WiFiManager(esp, secrets, status_light)
+pixels = neopixel.NeoPixel(board.D2, 16, brightness=1.0, auto_write=False)
+pixels.fill(0x050505)
+pixels.show()
+
+# clouds palette
+cloudy_palette = [fancy.CRGB(1.0, 1.0, 1.0), # White
+                  fancy.CRGB(0.5, 0.5, 0.5), # gray
+                  fancy.CRGB(0.5, 0.5, 1.0)] # blue-gray
+# sunny palette
+sunny_palette = [fancy.CRGB(1.0, 1.0, 1.0), # White
+                 fancy.CRGB(1.0, 1.0, 0.0),# Yellow
+                 fancy.CRGB(1.0, 0.5, 0.0),] # Orange
+# thunderstorm palette
+thunder_palette = [fancy.CRGB(0.0, 0.0, 1.0), # blue
+                   fancy.CRGB(0.5, 0.5, 0.5), # gray
+                   fancy.CRGB(0.5, 0.5, 1.0)] # blue-gray
+last_thunder_bolt = None
+
+palette = None  # current palette
+pal_offset = 0  # Positional offset into color palette to get it to 'spin'
+levels = (0.25, 0.3, 0.15)  # Color balance / brightness for gamma function
+raining = False
+thundering = False
+
+weather_refresh = None
+weather_type = None
+while True:
+    # only query the weather every 10 minutes (and on first run)
+    if (not weather_refresh) or (time.monotonic() - weather_refresh) > 600:
```

Comment:
Could you init `weather_refresh` to some negative number, like `-600` to make it run on first pass? Then this could just become:
```python
if time.monotonic() - weather_refresh > 600:
```

Reply:
i kinda prefer this cause you could 'force' refresh whenever you want :)